src/cleanup_guard.rs

```diff
@@ -0,0 +1,122 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Mutex;
+use std::{io, thread};
+
+use lazy_static::lazy_static;
+use slab::Slab;
+
+lazy_static! {
+    /// Contains the callbacks passed to currently-live [`CleanupGuard`]s
+    static ref LIVE_GUARDS: Mutex<GuardTable> = Mutex::new(Slab::new());
+}
+
+type GuardTable = Slab<Box<dyn FnOnce() + Send>>;
+
+/// Prepare to run [`CleanupGuard`]s on `SIGINT`/`SIGTERM`
+pub fn init() {
+    static CALLED: AtomicBool = AtomicBool::new(false);
+
+    if CALLED.swap(true, Ordering::Relaxed) {
+        return;
+    }
+    // Safety: check above ensures at most one call
+    if let Err(ref e) = unsafe { platform::init() } {
+        eprintln!("couldn't register signal handler: {}", e);
+    }
+}
+
+/// A drop guard that also runs on `SIGINT`/`SIGTERM`
+// Could be extended to not require `'static` at the cost of significantly more unsafe.
+pub struct CleanupGuard {
+    slot: usize,
+}
+
+impl CleanupGuard {
+    /// Invoke `f` when dropped or killed by `SIGINT`/`SIGTERM`
+    pub fn new<F: FnOnce() + Send + 'static>(f: F) -> Self {
+        let guards = &mut *LIVE_GUARDS.lock().unwrap();
+        Self {
+            slot: guards.insert(Box::new(f)),
+        }
+    }
+}
+
+impl Drop for CleanupGuard {
+    fn drop(&mut self) {
+        let guards = &mut *LIVE_GUARDS.lock().unwrap();
+        let f = guards.remove(self.slot);
+        f();
+    }
+}
+
+// Invoked on a background thread. Process exits after this returns.
+fn on_signal(guards: &mut GuardTable) {
+    for guard in guards.drain() {
+        guard();
+    }
+}
+
+#[cfg(unix)]
+mod platform {
```

Comment:
Have you seen [the `signal_hook` crate](https://docs.rs/signal-hook/0.3.14/signal_hook/)? It's quite frequently used (we already depend on it via `crossterm`) and it seems that it's meant to help with some of this (though I haven't checked the details).

Reply:
I reviewed it in some depth along with `ctrlc`. I'm reluctant to adopt it for two reasons:
- It seems heavily focused around mediating between multiple signal handlers. We don't need this.
- The "signal handler thread" pattern, along with most of the other high level safe helpers, appears to be only supported on Unix, so a portable approach would need most of the code written here anyway.

Reply:
That's a shame :( Thanks for checking!